(CEV2-266) Problem: Missing ClientRPC documentation UI (#208)
crypto-com/chain-docs: docs/api/client-rpc.md

# ClientRPC API Documentation

This documents lists all the RPC methods available on top of the `client-rpc` component.

## Methods

:::tip
All the methods explained in this document assume the following settings.
:::

```bash
Method: POST
URL: http://localhost:26659/ 
Headers: Content-Type: application/json
```
* JSONRPC requests can be `POST`'d to the root RPC endpoint via HTTP. Example below:
```bash
curl --header "Content-Type: application/json" --request POST --data '{"method": "<<method_name>>", "params": ["5"], "id": 1}' localhost:26659
```
* In the examples used here 
    - `RPC_HOST_NAME` = `localhost`
    - `RPC_SERVER_PORT` = `26659`

Remember to change your `RPC_HOST_NAME` and `RPC_SERVER_PORT` (if different used).

* If any call fails the response `body` contains an `error` Object: [ErrorResponse](./api-objects.md#errorresponse)
* For all the API-Objects used here, refer [this page](./api-objects.md).

> * To download available methods as a **POSTMAN** Collection, [click here](https://github.com/cdc-Hitesh/chain-docs/tree/master/docs/api/client-rpc-methods.postman_collection.json)

Author: I am not sure, Should I better keep it off the repo?

> * To download available methods as a **POSTMAN** Collection, [click here](https://github.com/cdc-Hitesh/chain-docs/tree/master/docs/api/client-rpc-methods.postman_collection.json)

Reviewer: It is better put the maintained collection in Crypto.com repository or hosted on Crypto.com website.

***Body:***

```json       
{
	"method": "wallet_createStakingAddress",
	"jsonrpc": "2.0",
	"params": [
        {
            "name": "{{wallet_name}}",
            "enckey": "{{wallet_enckey}}"
        }
    ],
	"id": "wallet_createStakingAddress"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": "0x78390b88c29a15dcb6fa0ba786063bfe4d3e74c0",
    "id": "wallet_create "
}
```

### `wallet_createStakingAddressBatch`

Create multiple staking address in a single call.

#### Parameters
* [WalletRequest](./api-objects.md#walletrequest) - *Object* - Wallet Authentication Object.
* *UInt32* - Number of staking address to create.

#### Returns
`result` : *UInt32* - Number of staking addresses created.

***Body:***

```json       
{
	"method": "wallet_createStakingAddressBatch",
	"jsonrpc": "2.0",
	"params": [
        {
    		"name": "{{wallet_name}}",
    		"enckey": "{{wallet_enckey}}"
    	},
         5
    ],
	"id": "wallet_createStakingAddressBatch"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": 5,
    "id": "wallet_createStakingAddressBatch"
}
```

### `wallet_createWatchStakingAddress`

> Create a Watch only Staking address.

Reviewer 2: ```suggestion
Imports a watch only staking address.
```

#### Parameters
* [WalletRequest](./api-objects.md#walletrequest) - *Object* - Wallet Authentication Object.

> * *String* - Public key used in Crypto.com Chain. [@ToDo - Any technical language to be put here?]

Reviewer 2: ```suggestion
* *String* - Secp256k1 public key of the staking address in either compressed or uncompressed form.
```

#### Returns
`result` : *String* - `Address` Generated Watch staking address.

***Body:***

```json       
{
	"method": "wallet_createWatchStakingAddress",
	"jsonrpc": "2.0",
	"params": [
        {
    		"name": "{{wallet_name}}",
    		"enckey": "{{wallet_enckey}}"
    	},
        "{{your_public_key}}"
    ],
	"id": "wallet_createWatchStakingAddress"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": "<<Generated Watch Staking addrss>>",
    "id": "wallet_createWatchStakingAddress"
}
```

### `wallet_createTransferAddress`

> @Todo - Do we need to give hyperlink to what does these terms mean?

Author: Yes yes. I get this. Let me refer this/

***Body:***

```json       
{
	"method": "wallet_listPublicKeys",
	"jsonrpc": "2.0",
	"params": [
        {
            "name": "{{wallet_name}}",
            "enckey": "{{wallet_enckey}}"
        }
    ],
	"id": "wallet_listPublicKeys"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": [
        "03ac4a6b1a14e848f...........699061ba5ef6c093ee30162fefc9317037d4a",
        "03a34bc9d9ffb53d5...........d352bccd0acabda957bc2a3e49a12499cd4b06"
    ],
    "id": "wallet_listPublicKeys"
}
```

### `wallet_listStakingAddresses`

Retrieve all the Staking addresses available.

#### Parameters
* [WalletRequest](./api-objects.md#walletrequest) - *Object* - Wallet Request Object.
* *UInt32* - Offset value (to indicate the `cursor`).
* *Uint32* - Maximum number of records to fetch.
* *Boolean* - if `true` returns in reverse-chronological order. //@Todo - wording needs to change or ok?

#### Returns
`result` : *Array(String)* - List of all Staking addresses.

***Body:***

```json       
{
	"method": "wallet_listStakingAddresses",
	"jsonrpc": "2.0",
	"params": [
        {
            "name": "{{wallet_name}}",
            "enckey": "{{wallet_enckey}}"
        }

>     //2, 100, false

Author: I kept it intentionally, though they are optional arguments actual RPC call wasn't accepting these parameters

            "value": "9999699998146"
        }
    ],
    "id": "wallet_listUTxO"
}
```

### `wallet_broadcastSignedTransferTx`

Broadcast a signed Transfer transaction to the blockchain.

#### Parameters
* [WalletRequest](./api-objects.md#walletrequest) - *Object* - Wallet Request Object.
* *String* - Signed raw transaction. //@Todo: need to specify something specific here ?

#### Returns
`result` : *String* - Transaction ID (Hex Encoded). // @Todo: tx_id needs to be hyperlinked ?

***Body:***

```json       
{
	"method": "wallet_broadcastSignedTransferTx",
	"jsonrpc": "2.0",
	"params": [
		{
			"name": "{{wallet_name}}",
			"enckey": "{{wallet_enckey}}"
		},
        "<<YOUR SIGNED HEX>>"
	],
	"id": "wallet_broadcastSignedTransferTx"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": "<<HEX ENCODED TX_ID>>",
    "id": "wallet_broadcastSignedTransferTx"
}
```

### `wallet_restore`
Restores a `HD` or `HW` type wallet. //@Todo: need hyperlink ?

#### Parameters
* [CreateWalletRequest](./api-objects.md#createwalletrequest) - *Object* - Wallet Request Object.
* *String* - valid Mnemonic phrase. (eg. `alpha beta charlie .. . . . .  . zeta`)

#### Returns

> `result` : *String* - Wallet Authentication Token or Secret Key. //@Todo : This is `enckey`, so hyperlink? 

Author: Great thanks.

***Body:***

```json       
{
    "method": "wallet_restore",
    "jsonrpc": "2.0",
    "params": [
        {
            "name": "{{wallet_name}}",
            "passphrase": "{{wallet_passphrase}}"
        },
        "ordinary mandate edit father snack mesh history identify print borrow skate unhappy cattle tiny first"
    ],
    "id": "wallet_restore"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": "<<Wallet Authentication Token>>", //@Todo
    "id": "wallet_restore"
}
```

### `wallet_restoreBasic`

> Restore a `Basic` type wallet. //@Todo - hyperlink needed to explain wallet types?

Author: Hyperlink this one for now

#### Returns
`result` : *null*

***Body:***

```json       
{
    "method": "wallet_delete",
    "jsonrpc": "2.0",
    "params": [
        {
            "name": "{{wallet_name}}",
            "passphrase": "{{wallet_passphrase}}"
        }
    ],
    "id": "wallet_delete"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": null,
    "id": "wallet_delete"
}
```

### `wallet_sendToAddress`
Send a `Transfer` type transaction from a wallet.

#### Parameters
* [WalletRequest](./api-objects.md#walletrequest) - *Object* - Wallet Request Object.
* *String* - Destination/To Address.
* *Uint64* - Amount to send in base unit (to be Stringified).
* *Array(String)* - (Optional) List of View Keys.

#### Returns
`result` : *String* - Transaction ID. 

***Body:***

```json       
{
	"method": "wallet_sendToAddress",
	"jsonrpc": "2.0",
	"params": [
		{
			"name": "{{wallet_name}}",
			"enckey": "{{wallet_enckey}}"
		},
		"dcro1ayhu0665wprxf86letqlv8x4ssppeu6awf7m60qlwds9268vltwsk6ehwa",
		"125",
		[]
	],
	"id": "wallet_sendToAddress"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",

>     "result": "<<tx_id>>",

Reviewer 2: ```suggestion
    "result": "<<TX_ID>>",
```

Reviewer 2: @cdc-Hitesh

        "key_chainpath": "AA==",
        "key_pairs": "CAUBBKNLyL1kIgljmzt7gDZ7fEsfZtFYHnvXZ0iJNCzrvUXdHQXDH6HyPjjX0aUq",
        "multisig_address_pair": "CAEBNjU0NDQ5Yjdnz1GZ+/dMpR7jTlL5KCCjS8nZ/7U9WtSYXxB201K8zQrKvalXvCo+SaEkmc1LBgEAAAAAAAAAAAAAAAAAAAA=",
        "name": "test_account",
        "passphrase": null,
        "private_key": "gHvvwqvI7mV0VbkiLNhXZ6efE",
        "wallet": "BQEETqp27JejrjeIjRUPkeH9VuTlfT4S2gX6lBN9F50TdaWraOtT0NJGhskelHjVdVmfPUZn790ylHuNOUvkoIAE="
    },
    "id": "wallet_export"
}
```

### `wallet_importTransaction`
Imports a Chain transaction exported from [wallet_exportTransaction](#wallet_exportTransaction).

#### Parameters
* [WalletRequest](./api-objects.md#walletrequest) - *Object* - Wallet Request Object.
* *String* - encoded raw transaction.

#### Returns
`result` : *Quantity*, String - Amount of the transaction.

***Body:***

```json       
{
	"method": "wallet_importTransaction",
	"jsonrpc": "2.0",
	"params": [
		{
			"name": "{{wallet_name}}",
			"enckey": "{{wallet_enckey}}"
        }
        , "encoded_tx_goes_here"
	],
	"id": "wallet_importTransaction"
}
```

> @ToDo

Reviewer 2: @cdc-Hitesh

***Body:***

```json       
{
	"method": "staking_depositAmountStake",
	"jsonrpc": "2.0",
	"params": [
		{
			"name": "{{wallet_name}}",
			"enckey": "{{wallet_enckey}}"
		},
		"{{recipient_address}}",
		"500000000000000000"
	],
	"id": "staking_depositAmountStake"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": "<<tx_id>>",
    "id": "staking_depositAmountStake"
}
```

###  `staking_withdrawAllUnbondedStake`

Withdraws all unbonded tokens to a destination transfer address.

#### Parameters
* *Object* - [WalletRequest](./api-objects.md#walletrequest) - Wallet authentication object.
* *String* - `from` Address.
* *String* - `to` Address.
* *Array* - (Optional) List of view keys.

#### Returns
`result` : *String* - Transaction ID of the blockchain submitted transaction.

***Body:***

```json       
{
	"method": "staking_withdrawAllUnbondedStake",
	"jsonrpc": "2.0",
	"params": [
		{
			"name": "{{wallet_name}}",
			"enckey": "{{wallet_enckey}}"
		},
		"{{from_address}}",
		"{{to_address}}",
		[ "{{Optional_view_key_1}}",  "{{Optional_view_key_1}}" ]
	],
	"id": "staking_withdrawAllUnbondedStake"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",

>     "result": "<<tx_id>>",

Reviewer 2: ```suggestion
    "result": "<<TX_ID>>",
```

    "id": "staking_withdrawAllUnbondedStake"
}
```

###  `staking_depositStake`
Deposit staked tokens to an address.

#### Parameters
* *Object* - [WalletRequest](./api-objects.md#walletrequest) - Wallet authentication object.
* *String* - `to` Address.
* *Array([TxoPointer](./api-objects.md#txopointer))* - List of UTXOs.

#### Returns
`result` : *String* - Transaction ID of the blockchain submitted transaction.

***Body:***

```json       
{
	"method": "staking_depositStake",
	"jsonrpc": "2.0",
	"params": [
		{
			"name": "{{wallet_name}}",
			"enckey": "{{wallet_enckey}}"
		},
		"{{staking_address}}",
		[
			{
				"id": "5c40805013482d6c293a09b0f6aa532cbd338bf127c0954f1960d608262b5115",
				"index": 0
			}
		]
	],
	"id": "staking_depositStake"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",

>     "result": "<<tx_id>>",

Reviewer 2: ```suggestion
    "result": "<<TX_ID>>",
```

{
    "method": "multiSig_addNonce",
    "jsonrpc": "2.0",
    "params": [
    	"{{session_id}}",
        "{{wallet_enckey or SecKey}}",
        "{{nonce_value}}",
        "02b2a02cb8eacc6a0602a976b6a2dcd6924b68120b42c9843cdd6191606d0a1350"
    ],
    "id": "multiSig_addNonce"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": null,
    "id": "multiSig_addNonce"
}
```

### `multiSig_addNonceCommitment`

Commits a nonce for a multisig transaction.

#### Parameters
* *String* - Session ID retrieved after creating a new session.
* *String* - Wallet authentication key.
* *String* - Value retrieved from [multiSig_nonceCommitment](#multiSig_nonceCommitment) 
* *String* - Signer's public key
 
#### Returns
`result` : *Object* - `null`

***Body:***

```json       
{
    "method": "multiSig_addNonceCommitment",
    "jsonrpc": "2.0",
    "params": [
    	"{{session_id}}",
        "{{wallet_enckey}}",
        "{{nonce commitment value}}",
        "{{merchant_public_key}}"
    ],
    "id": "multiSig_addNonceCommitment"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": null,
    "id": "multiSig_addNonceCommitment"
}
```

### `multiSig_createAddress`

> It can be used to create a `m-of-n` multisig wallet. Where `m` is the minimum number of signatures required, `n` is the total number of parties.

Reviewer 3: Can add sth like "n has to be greater or equal to m"

Author: Okay good one i will do it

    ],
    "id": "multiSig_listAddressPublicKeys"
}
```

### `multiSig_newSession`

Create a new session ID for a multiSig transaction.

#### Parameters
* [WalletRequest](./api-objects.md#walletrequest) - *Object* - Wallet Authentication Object.
* *String* - Message to pass
* *Array(String)* - An array of valid participating public keys
* *String* - Signer's public key

#### Returns
`result` : *String* - session id (To be used as `mulisig_session_id`).

***Body:***

```json       
{
    "method": "multiSig_newSession",
    "jsonrpc": "2.0",
    "params": [
        {
            "name": "{{wallet_name}}",
            "enckey": "{{wallet_enckey}}"
        },
        "{{transaction_message}}",
        [
        	"{{merchant_public_key}}",
            "{{public_key}}"
        ],
        "{{public_key}}"
    ],
    "id": "multiSig_newSession"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": "{{multisig_session_id}}",
    "id": "multiSig_newSession"
}
```

### `multiSig_nonce`
Requests a new nonce for the multisig transaction.

#### Parameters
* *String* - Session ID retrieved after creating a new session.
* *String* - Wallet authentication key.
 
#### Returns

> `result` : *String* - Serialised nonce value.

Reviewer 2: @cdc-Hitesh

***Body:***

```json       
{
    "method": "multiSig_partialSign",
    "jsonrpc": "2.0",
    "params": [
    	"{{multisig_session_id}}",
        "{{wallet_enc_key}}"
    ],
    "id": "multiSig_partialSign"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": "<<serialised_partial_sign",
    "id": "multiSig_partialSign"
}
```

### `multiSig_addPartialSignature`

Adds a partial signature to a multisig_transaction.

#### Parameters
* *String* - Session ID retrieved after creating a new session.
* *String* - Wallet authentication key.
* *String* - Partial signature value retrieved from [multiSig_partialSign](#multiSig_partialSign) 
* *String* - Signer's public key
 
#### Returns
`result` : *Object* - `null`.

***Body:***

```json       
{
    "method": "multiSig_addPartialSignature",
    "jsonrpc": "2.0",
    "params": [
    	"{{session_id}}",
        "{{wallet_enckey}}",
        "{{partial signature value}}",
        "{{signer public key}}"
    ],
    "id": "multiSig_addPartialSignature"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": null,
    "id": "multiSig_addPartialSignature"
}
```

### `multiSig_signature`

> Start applying the multisig signature process.

Reviewer 2: ```suggestion
Returns the final signature of an existing multiSig transaction session.
```

#### Parameters
* *String* - Session ID retrieved from [multiSig_newSession](#multiSig_newSession).
* *String* - Wallet authentication key.
 
#### Returns
`result` : *String* - multisig_Signature value.

***Body:***

```json       
{
    "method": "multiSig_signature",
    "jsonrpc": "2.0",
    "params": [
    	"{{session_id}}",
        "{{wallet_enckey}}"
    ],
    "id": "multiSig_signature"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": "<<serialised multisig signature initiation>>",
    "id": "multiSig_signature"
}
```

### `multiSig_broadcastWithSignature`

> Signs a provided unsigned transaction and broadcast it to the blockchain.

Reviewer 2: ```suggestion
Creates and attaches the final signature of an existing multiSig transaction session to the corresponding unsigned transaction provided and broadcast it to the blockchain.
```

#### Parameters
* *Object* - [WalletRequest](./api-objects.md##WalletRequest)
* *String* - Session ID retrieved from [multiSig_newSession](#multiSig_newSession).
* *Object* - [Tx](./api-objects.md#tx) - Unsigned TX object. //@todo: is it ok?
 
#### Returns
`result` : *String* - `tx_id` or Transaction ID after submission.

> ***Body: //@Todo : complete body below***

Reviewer 2: Resolve this when this is done.

```json       
{
    "method": "multiSig_broadcastWithSignature",
    "jsonrpc": "2.0",
    "params": [
    	"{{session_id}}",
        "{{wallet_enckey}}"
    ],
    "id": "multiSig_broadcastWithSignature"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",

>     "result": "<<Tx_ID>>",

Reviewer 2: ```suggestion
    "result": "<<TX_ID>>",
```

    "id": "multiSig_broadcastWithSignature"
}
```

## TRANSACTION CREATOR

### `transaction_createRaw`

Creates a unsigned [RawTransaction](./api-objects.md#rawtransaction) with the inputs provided.

#### Parameters
* *Array([TxoPointer](./api-objects.md#txopointer))* - UTXO  inputs.
* *Array([TxOut](./api-objects.md#txout))* - Expected Outputs.
* *Array* - List of View keys to be used.
 
#### Returns
`result` : *Object* - [RawTransaction](./api-objects.md#rawtransaction)

> ***@Todo :Complete body below Body:***

Reviewer 2: Resolve this when this is done

```json  
{
    "method": "transaction_createRaw",
    "jsonrpc": "2.0",
    "params": [
    	"{{session_id}}",
        "{{wallet_enckey}}",
        "02b2a02cb8eacc6a0602a976b6a2dcd6924b68120b42c9843cdd6191606d0a1350",
        "{{merchant_public_key}}"
    ],
    "id": "transaction_createRaw"
}
```

> //@todo: complete response

Reviewer 2: Resolve this when this is done

            "moniker": "node0",
            "network": "test-chain-y3m1e6-AB",
            "other": {
                "rpc_address": "tcp://0.0.0.0:26657",
                "tx_index": "on"
            },
            "protocol_version": {
                "app": "1",
                "block": "10",
                "p2p": "7"
            },
            "version": "0.33.4"
        },
        "sync_info": {
            "catching_up": false,
            "latest_app_hash": "2D742A6764C4E9109F520B27CED5DEB98D7656222E8317C30D19DE3E21265AC9",
            "latest_block_hash": "3AE3E08FB7E2A09E747073BF9AFC27DA8AFA466446B27BA85F7927AC0C39CBA8",
            "latest_block_height": "556696",
            "latest_block_time": "2020-07-27T11:04:19.779027255Z"
        },
        "validator_info": {
            "address": "0D8912E96DAAE332E7393EC1DC4BA2D249A44445",
            "proposer_priority": null,
            "pub_key": {
                "type": "tendermint/PubKeyEd25519",
                "value": "FF5JxhRrCUNLj6UZmYdjv/AWgSWUeiomeOMeJG71owE="
            },
            "voting_power": "5000000000"
        }
    },
    "id": "status"
}
    
```

### `genesis`

> Returns the Genesis information of the network.

Author: Ok done.

***Body:***

```json       
{
    "method": "wallet_restoreBasic",
    "jsonrpc": "2.0",
    "params": [
        {
            "name": "{{wallet_name}}",
            "passphrase": "{{wallet_passphrase}}"
        },
        "<< Wallet Private Key >>"
    ],
    "id": "wallet_restoreBasic"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": "<<Wallet Authentication Token>>",
    "id": "wallet_restoreBasic"
}
```

### `wallet_delete`
Delete a wallet. 

#### Parameters
* [CreateWalletRequest](./api-objects.md#createwalletrequest) - *Object* - CreateWallet Request Object.

#### Returns
`result` : *null*

***Body:***

```json       
{
    "method": "wallet_delete",
    "jsonrpc": "2.0",
    "params": [
        {
            "name": "{{wallet_name}}",
            "passphrase": "{{wallet_passphrase}}"
        }
    ],
    "id": "wallet_delete"
}
```

***Response Body***
```json
{
    "jsonrpc": "2.0",
    "result": null,
    "id": "wallet_delete"
}
```

### `wallet_sendToAddress`
Send a `Transfer` type transaction from a wallet.

#### Parameters
* [WalletRequest](./api-objects.md#walletrequest) - *Object* - Wallet Request Object.

> * *String* - Destination/To Address.

Author: Thanks for pointing out.
